share/availability/options.go

```diff
@@ -0,0 +1,344 @@
+package availability
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/celestiaorg/celestia-node/share"
+)
+
+var (
+	// ErrInvalidAvailOption is an error that is returned by FullAvailabilityPameters.Validate
+	// LightAvailabilityPameters.Validate and ShareParameters.Validate when supplied with invalid values.
+	// This error will also be returned by NewShareAvailability if supplied with an invalid option
+	ErrInvalidAvailOption = fmt.Errorf("availability: invalid option")
+
+	// ErrInvalidAvailOption is an error that is returned by DiscoveryParameters.Validate
+	// and ShareParameters.Validate when supplied with invalid values.
+	// This error will also be returned by NewDiscovery if supplied with an invalid option
+	ErrInvalidDiscOption = fmt.Errorf("discovery: invalid option")
+)
+
+// ErrInvalidOption is a utility function to dedup code for error-returning
+// when dealing with invalid parameter values.
+func ErrInvalidOption(err error, optionName string, value string) error {
+	return fmt.Errorf("%w: value %s cannot be %s", err, optionName, value)
+}
+
+// Parameters is an interface tjat encompasses all params needed for the share package
+type Parameters interface {
+	FullAvailabilityParameters | LightAvailabilityParameters | CacheAvailabilityParamaters | DiscoveryParameters
+}
+
+// Option is a function that configures a ShareParameters
+type Option[T Parameters] func(*T)
+
+// FullAvailabilityParameters is the set of parameters that must be configured for the full availability implementation
+type FullAvailabilityParameters struct {
+	// AvailabilityTimeout specifies timeout for DA validation during which data have to be found on the network,
+	// otherwise ErrNotAvailable is fired.
+	// TODO: https://github.com/celestiaorg/celestia-node/issues/10)
+	AvailabilityTimeout time.Duration
+}
```

Comment:
Lets put specific params into corresponding subpackages.

Reply:
Putting them in their corresponding subpackages and then importing them in `share/availability/options.go` to aggregate all parameters for configuration's sake would cause import cycles. I suggest we keep it as is.

Reply:
You can make options typed and pkg specific (not generic I know). Subpkg should be fine importing higher level pkg. Otherwise I don't see where is import cycle

Reply:
Okay we can sync and I can show you.